L1Trigger/DTTriggerPhase2/src/MuonPathAssociator.cc

```diff
@@ -0,0 +1,1028 @@
+
+using namespace edm;
+using namespace std;
+
+// ============================================================================
+// Constructors and destructor
+// ============================================================================
+MuonPathAssociator::MuonPathAssociator(const ParameterSet &pset, edm::ConsumesCollector &iC) {
+  // Obtention of parameters
+  debug_ = pset.getUntrackedParameter<bool>("debug");
+  clean_chi2_correlation_ = pset.getUntrackedParameter<bool>("clean_chi2_correlation");
+  use_LSB_ = pset.getUntrackedParameter<bool>("use_LSB");
+  tanPsi_precision_ = pset.getUntrackedParameter<double>("tanPsi_precision");
+  x_precision_ = pset.getUntrackedParameter<double>("x_precision");
+  useBX_correlation_ = pset.getUntrackedParameter<bool>("useBX_correlation");
+  allow_confirmation_ = pset.getUntrackedParameter<bool>("allow_confirmation");
+  dT0_correlate_TP_ = pset.getUntrackedParameter<double>("dT0_correlate_TP");
+  dBX_correlate_TP_ = pset.getUntrackedParameter<int>("dBX_correlate_TP");
+  dTanPsi_correlate_TP_ = pset.getUntrackedParameter<double>("dTanPsi_correlate_TP");
+  minx_match_2digis_ = pset.getUntrackedParameter<double>("minx_match_2digis");
+  chi2corTh_ = pset.getUntrackedParameter<double>("chi2corTh");
+
+  if (debug_)
+    LogDebug("MuonPathAssociator") << "MuonPathAssociator: constructor";
+
+  //shift
+  int rawId;
+  shift_filename_ = pset.getParameter<edm::FileInPath>("shift_filename");
+  std::ifstream ifin3(shift_filename_.fullPath());
+  double shift;
+  if (ifin3.fail()) {
+    throw cms::Exception("Missing Input File")
+        << "MuonPathAnalyzerPerSL::MuonPathAnalyzerPerSL() -  Cannot find " << shift_filename_.fullPath();
+  }
+  while (ifin3.good()) {
+    ifin3 >> rawId >> shift;
+    shiftinfo_[rawId] = shift;
+  }
+
+  dtGeomH_ = iC.esConsumes<DTGeometry, MuonGeometryRecord, edm::Transition::BeginRun>();
+}
+
+MuonPathAssociator::~MuonPathAssociator() {
+  if (debug_)
+    LogDebug("MuonPathAssociator") << "MuonPathAssociator: destructor";
+}
+
+// ============================================================================
+// Main methods (initialise, run, finish)
+// ============================================================================
+void MuonPathAssociator::initialise(const edm::EventSetup &iEventSetup) {
+  if (debug_)
+    LogDebug("MuonPathAssociator") << "MuonPathAssociator::initialiase";
+
+  const MuonGeometryRecord &geom = iEventSetup.get<MuonGeometryRecord>();
+  dtGeo_ = &geom.get(dtGeomH_);
+}
+
+void MuonPathAssociator::run(edm::Event &iEvent,
+                             const edm::EventSetup &iEventSetup,
+                             edm::Handle<DTDigiCollection> digis,
+                             std::vector<metaPrimitive> &inMPaths,
+                             std::vector<metaPrimitive> &outMPaths) {
+  if (dT0_correlate_TP_)
+    correlateMPaths(digis, inMPaths, outMPaths);
+  else {
+    for (auto metaPrimitiveIt : inMPaths)
```

Comment:
replace all of this with `outMPaths.insert(outMPaths.end(),inMPaths.begin(),inMPaths.end());`